include/circt/Dialect/Ibis/IbisOps.td

```diff
@@ -145,4 +160,110 @@ def CallOp : IbisOp<"call", [CallOpInterface]> {
   }];
 }
 
+
+// ===---------------------------------------------------------------------===//
+// Low-level Ibis operations
+// ===---------------------------------------------------------------------===//
+
+def ContainerOp : IbisOp<"container", [
+  IsolatedFromAbove, RegionKindInterface,
+  Symbol, SymbolTable, SingleBlock,
+  NoTerminator, NoRegionArguments,
+  HasParent<"ClassOp">
+]> {
+  let summary = "Ibis container";
+  let description = [{
+    An ibis container describes a collection of logic nested within an Ibis class.
+  }];
+
+  let arguments = (ins SymbolNameAttr:$sym_name);
+  let regions = (region SizedRegion<1>:$body);
+
+  let assemblyFormat = [{
+    $sym_name attr-dict-with-keyword $body
+  }];
+
+  let extraClassDeclaration = [{
+    // Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { return RegionKind::Graph; }
+  }];
+}
+
+class PortLikeOp<string mnemonic, list<Trait> traits = []> :
+    IbisOp<mnemonic, !listconcat(traits, [
+      PortOpInterface,
+      HasParent<"ClassOp">
+  ])> {
+  let arguments = (ins SymbolNameAttr:$sym_name, TypeAttrOf<AnyType>:$type);
+  let assemblyFormat = [{
+    $sym_name `:` $type attr-dict
+  }];
+}
+
+def InputPortOp : PortLikeOp<"port.input"> {
+  let summary = "Ibis input port";
+}
+
+def OutputPortOp : PortLikeOp<"port.output"> {
+  let summary = "Ibis output port";
+}
+
+def PortReadOp : IbisOp<"port.read", [
+      DeclareOpInterfaceMethods<SymbolUserOpInterface>
+]> {
+  let summary = "Ibis port read";
+  let description = [{
+    Read the value of a local port (i.e. a port of 'this' class).
+  }];
+
+  let arguments = (ins FlatSymbolRefAttr:$sym_name);
+  let results = (outs AnyType:$output);
+  let assemblyFormat = [{
+    $sym_name attr-dict `:` type($output)
+  }];
+}
+
+def PortWriteOp : IbisOp<"port.write", [
+      DeclareOpInterfaceMethods<SymbolUserOpInterface>
+]> {
+  let summary = "Ibis port write";
+  let description = [{
+    Write a value to a local port (i.e. a port of 'this' class).
+  }];
+
+  let arguments = (ins FlatSymbolRefAttr:$sym_name, AnyType:$input);
+  let assemblyFormat = [{
+    $sym_name `(` $input `)` attr-dict `:` type($input)
+  }];
+}
+
+def InstanceReadOp : IbisOp<"instance.read", [
```

Comment:
I would suggest the mnemonic "instance.port.read" (and similar for write) to be clear about it.